### PR TITLE
Update_prognostic_ice subroutine port

### PR DESCRIPTION
**Pull request by singhbalwinder**
Ports "update_prognostic_ice" subroutine of P3 in C++/Kokkos with a BFB unit test.

### CODE REVIEW

**Comment by singhbalwinder** [comment]
I am requesting a review as this is my first attempt (with HUGE amount of help from Kyle, Jim and Andrew B.!). I have not tested it on GPUs, I will do that in the meanwhile.

File: components/scream/src/physics/p3/p3_functions_f90.cpp [hunk 1]
```diff
@@ -534,6 +552,79 @@ void get_rain_dsd2_f(Real qr_, Real* nr_, Real* mu_r_, Real* lamr_, Real* cdistr
...
+  Real local_qc	    = *qc_;	  
+  Real local_nc	    = *nc_;
+  Real local_qr	    = *qr_;	  
+  Real local_nr	    = *nr_;	  
+  Real local_qitot  = *qitot_;
+  Real local_nitot  = *nitot_;
+  Real local_qirim  = *qirim_;
+  Real local_birim  = *birim_;
+
+
```
Comment (jgfouca) [inline components/scream/src/physics/p3/p3_functions_f90.cpp]: Nitpick: only one blank line is needed to separate code blocks.

File: components/scream/src/physics/p3/p3_functions_f90.hpp [hunk 2]
```diff
@@ -394,6 +394,64 @@ void calc_bulk_rho_rime_f(Real qi_tot, Real* qi_rim, Real* bi_rim, Real* rho_rim
 
 }
 
+///////////////////////////////////////////////////////////////////////////////
+
+struct P3UpdatePrognosticIceData
+{
+  //inputs
+  Real qcheti;
```
Comment (jgfouca) [inline components/scream/src/physics/p3/p3_functions_f90.hpp]: I would compress this:
```
Real qcheti, qccol, qcshd ...
```

File: components/scream/src/physics/p3/p3_functions_update_prognostics.cpp [hunk 3]
```diff
@@ -0,0 +1,15 @@
+#include "p3_functions_update_prognostics_impl.hpp"
+#include "share/scream_types.hpp"
+
+namespace scream {
+  namespace p3 {
```
Comment (jgfouca) [inline components/scream/src/physics/p3/p3_functions_update_prognostics.cpp]: Nitpick: in most of scream, we do not have namespaces increase the indent level.

File: components/scream/src/physics/p3/p3_functions_update_prognostics_impl.hpp [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    const Spack& nrshdr, const Spack& qimlt,  const Spack& nimlt,  const Spack& qisub,
+    const Spack& qidep,  const Spack& qinuc,  const Spack& ninuc,  const Spack& nislf,
+    const Spack& nisub,  const Spack& qiberg, const Spack& exner,  const Spack& xxls,
+    const Spack& xlf,    const bool log_predictNc, const bool log_wetgrowth, const Scalar dt,
+    const Spack& nmltratio, const Spack& rhorime_c, Spack& th, Spack& qv, Spack& qitot,
+    Spack& nitot, Spack& qirim, Spack& birim, Spack& qc,  Spack& nc, Spack& qr,
+    Spack& nr)
+{
+
+qc = qc - (qcheti + qccol + qcshd + qiberg) * dt;
```
Comment (jgfouca) [inline components/scream/src/physics/p3/p3_functions_update_prognostics_impl.hpp]: All the lines in this function need to be indented better. Emacs can auto-indent code blocks if that's what you're using.

File: components/scream/src/physics/p3/tests/p3_unit_tests.cpp [hunk 5]
```diff
@@ -201,9 +201,353 @@ static void  cloud_water_autoconversion_unit_bfb_tests(){
...
+     
+    KOKKOS_FUNCTION static void  update_prognostic_ice_unit_bfb_tests(){
+      using KTH = KokkosTypes<HostDevice>;
+      
+      static constexpr Int max_pack_size = 16;
+      
+      REQUIRE(Spack::n <= max_pack_size);
+      
+      //fortran generated data is input to the following 
+      // Jim's print statement for fortran code
```
Comment (jgfouca) [inline components/scream/src/physics/p3/tests/p3_unit_tests.cpp]: You can remove this. And I've been using ES10.4 for unit tests input where I don't care about precision as much.

File: components/scream/src/physics/p3/tests/p3_unit_tests.cpp [hunk 6]
```diff
@@ -201,9 +201,353 @@ static void  cloud_water_autoconversion_unit_bfb_tests(){
...
+           0.000000000000E+00,  0.000000000000E+00,  1.764574860854E-02,  0.000000000000E+00,  4.589050692518E-10,
+           1.085256336995E+00,  3.337000000000E+05,  2.834700000000E+06,  true,                true,
+           1.800000000000E+03,  2.000000000000E-01,  2.457038651196E+02,  2.848881406331E+02,  5.000000000000E-03,
+           8.571428571429E-05,  1.234478699837E+08,  8.947368421053E-06,  1.000000000000E+06,  1.000000000000E-04,
+           1.000000000000E+06,  8.571428571429E-05,  1.000000000000E-02},
+      };
+      
+      // Sync to device
+      KTH::view_1d<P3UpdatePrognosticIceData> pupidc_host("pupidc_host", Spack::n);
+      view_1d<P3UpdatePrognosticIceData> pupidc_device("pupidc_host", Spack::n);
```
Comment (jgfouca) [inline components/scream/src/physics/p3/tests/p3_unit_tests.cpp]: You should use Kokkos::create_mirror_view to get the host view, not KTH.

File: components/scream/src/physics/p3/tests/p3_unit_tests.cpp [hunk 7]
```diff
@@ -201,9 +201,353 @@ static void  cloud_water_autoconversion_unit_bfb_tests(){
...
+      
+      // This copy only copies the input variables.
+      std::copy(&pupidc[0], &pupidc[0] + Spack::n, pupidc_host.data());
+      Kokkos::deep_copy(pupidc_device, pupidc_host);
+
+      // Get data from fortran
+      for (Int i = 0; i < max_pack_size; ++i) {
+	update_prognostic_ice(pupidc[i]);
```
Comment (jgfouca) [inline components/scream/src/physics/p3/tests/p3_unit_tests.cpp]: strange not to increase indent level for a loop.

File: components/scream/src/physics/p3/tests/p3_unit_tests.cpp [hunk 8]
```diff
@@ -201,9 +201,353 @@ static void  cloud_water_autoconversion_unit_bfb_tests(){
...
+      }
+
+      // This copy also copies the output from the fortran function into the host view. These values
+      // are need to check the values returned from
+      std::copy(&pupidc[0], &pupidc[0] + Spack::n, pupidc_host.data());
```
Comment (jgfouca) [inline components/scream/src/physics/p3/tests/p3_unit_tests.cpp]: This copy is not needed. You can just use pupidc for validating CXX data.

File: components/scream/src/physics/p3/p3_functions_f90.hpp [hunk 9]
```diff
@@ -394,6 +394,64 @@ void calc_bulk_rho_rime_f(Real qi_tot, Real* qi_rim, Real* bi_rim, Real* rho_rim
...
+  Real exner;
+  Real xxls;
+  Real xlf;
+  bool log_predictNc;
+  bool log_wetgrowth;
+  Real dt;
+  Real nmltratio;
+  Real rhorime_c;
+
+  //output
```
Comment (jgfouca) [inline components/scream/src/physics/p3/p3_functions_f90.hpp]: These are inouts, not ouputs.

File: components/scream/src/physics/p3/tests/p3_unit_tests.cpp [hunk 10]
```diff
@@ -201,9 +201,353 @@ static void  cloud_water_autoconversion_unit_bfb_tests(){
...
+
+      // Validate results
+      //First verify the single value variables and then the ones in a pack
+      REQUIRE(pupidc[0].dt            == pupidc_host(0).dt);
+      REQUIRE(pupidc[0].log_predictNc == pupidc_host(0).log_predictNc);
+      REQUIRE(pupidc[0].log_wetgrowth == pupidc_host(0).log_wetgrowth);
+
+      for (Int s = 0; s < Spack::n; ++s) {
+
+	REQUIRE(pupidc[s].qcheti== pupidc_host(s).qcheti);
```
Comment (jgfouca) [inline components/scream/src/physics/p3/tests/p3_unit_tests.cpp]: There's no need to check inputs.

File: components/scream/src/physics/p3/tests/p3_unit_tests.cpp [hunk 11]
```diff
@@ -201,9 +201,353 @@ static void  cloud_water_autoconversion_unit_bfb_tests(){
...
+	  }
+	  
+	});
+      
+      // Sync back to host
+      Kokkos::deep_copy(pupidc_host, pupidc_device);
+
+      // Validate results
+      //First verify the single value variables and then the ones in a pack
+      REQUIRE(pupidc[0].dt            == pupidc_host(0).dt);
```
Comment (jgfouca) [inline components/scream/src/physics/p3/tests/p3_unit_tests.cpp]: Don't need to check these.

File: components/scream/src/physics/p3/tests/p3_unit_tests.cpp (hunk 6 above)
Comment (singhbalwinder) [reply components/scream/src/physics/p3/tests/p3_unit_tests.cpp]: I tried doing it but then the test started to fail. Would you point me to an example code which I can follow to address this?

File: components/scream/src/physics/p3/tests/p3_unit_tests.cpp (hunk 7 above)
Comment (singhbalwinder) [reply components/scream/src/physics/p3/tests/p3_unit_tests.cpp]: It looks indented in my code. I use Emacs.

File: components/scream/src/physics/p3/tests/p3_unit_tests.cpp (hunk 8 above)
Comment (singhbalwinder) [reply components/scream/src/physics/p3/tests/p3_unit_tests.cpp]: I will resolve this after making changes for "Kokkos::create_mirror_view"

File: components/scream/src/physics/p3/tests/p3_unit_tests.cpp (hunk 6 above)
Comment (jgfouca) [reply components/scream/src/physics/p3/tests/p3_unit_tests.cpp]:
```
view_1d<P3UpdatePrognosticIceData> pupidc_device("pupidc", Spack::n);
auto pupidc_host = Kokkos::create_mirror_view(pupidc_device);
```

Comment (singhbalwinder) [reply components/scream/src/physics/p3/tests/p3_unit_tests.cpp]: thanks!

**Comment by singhbalwinder** [comment]
@jgfouca : Thanks for the quick review and suggestions. I have now resolved all your comments. Please let me know if I missed/misunderstood anything.

File: components/scream/src/physics/p3/tests/p3_unit_tests.cpp [hunk 12]
```diff
@@ -201,9 +201,353 @@ static void  cloud_water_autoconversion_unit_bfb_tests(){
...
+  struct UnitWrap::UnitTest<D>::TestP3UpdatePrognosticIce
+  {
+  
+     
+    KOKKOS_FUNCTION static void  update_prognostic_ice_unit_bfb_tests(){
+      using KTH = KokkosTypes<HostDevice>;
+      
+      static constexpr Int max_pack_size = 16;
+      
+      REQUIRE(Spack::n <= max_pack_size);
```
Comment (bartgol) [inline components/scream/src/physics/p3/tests/p3_unit_tests.cpp]: You cannot use these macros inside a kokkos kernel. You need to do all the checks on host. That's because these macros can throw exceptions, and you can't do that in cuda.

Comment (bartgol) [reply components/scream/src/physics/p3/tests/p3_unit_tests.cpp]: If you look at all the other codes, we do the calculations, then deep_copy the results to host, and then check the results there.

Also, for things like the check above, you could use `assert(blah)`. Asserts are fine on cuda, and this looks like a pre-condition that should be met. The `REQUIRE` macro is to check tests results.

Comment (bartgol) [reply components/scream/src/physics/p3/tests/p3_unit_tests.cpp]: Wait, this is not a kernel though, right? If so, you should remove thee `KOKKOS_FUNCTION` macro. That triggers cuda compilation on GPU, which is not compatible with exceptions.

Comment (jgfouca) [reply components/scream/src/physics/p3/tests/p3_unit_tests.cpp]: @bartgol , i think this is not inside a kernel. @singhbalwinder did mistakenly put KOKKOS_FUNCTION on this function though.

Comment (bartgol) [reply components/scream/src/physics/p3/tests/p3_unit_tests.cpp]: Easy fix then. This is what caused all the white builds to fail. At least part of it.

Comment (singhbalwinder) [reply components/scream/src/physics/p3/tests/p3_unit_tests.cpp]: Thanks @bartgol for look at the code. I tried to follow other codes in that file and all functions have `KOKKOS_FUNCTION` . Should I remove that for all of them? Let me know if I misunderstood anything.

Comment (bartgol) [reply components/scream/src/physics/p3/tests/p3_unit_tests.cpp]: KOKKOS_FUNCTION is a macro that expands to `__host__ __device__` when the exec space is Cuda, and expands to nothing otherwhise. `KOKKOS_INLINE_FUNCTION` is similar, but in addition adds the `inline` keyword.

When a funciton is marked as `__device__`, it must be executable in a Cuda kernel. Part of the requirements is that such function cannot call functions that are not marked as `__device__` or throwing exceptions.

So, in short, if a function has to be executed inside a kernel, mark it as `KOKKOS_FUNCTION` or `KOKKOS_INLINE_FUNCTION`. If it's not supposed to be executed on device, do not mark it with any of those macros.

Comment (singhbalwinder) [reply components/scream/src/physics/p3/tests/p3_unit_tests.cpp]: Thanks @bartgol . I have removed that macro now.

**Comment by singhbalwinder** [comment]
I have fixed the broken test and this PR should be ready now.

**Comment by bartgol** [comment]
@singhbalwinder I think you need to merge master into the branch, so that the PR is deemed 'mergeable', and the AT can test it. Also, we cannot merge it (from github interface) unless conflicts are resolved, so we need a master merge.

**Comment by singhbalwinder** [comment]
Would it be okay if I rebase this branch off of the current master?

**Comment by bartgol** [comment]
In theory, we like to avoid rewriting git history once the PR is on. I think the default merge of PR on github is to create a merge commit, so rebasing on master rather than merging master in the PR would be a slightly different tree, but morally the same (you'd still see the 'branch' on the side of master, except that master would look as if without any commit while the PR was outstanding).

That said, I think the best thing to do is merge.

**Comment by singhbalwinder** [comment]
Thanks @bartgol ! I have now merged master into this branch and resolved all conflicts. All the tests are passing on my end so hopefully they will all pass with the auto-tester as well.
